File: 2-d Array/542. 01 Matrix.js
`Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.

The distance between two adjacent cells is 1.

 

Example 1:


Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]
Example 2:


Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]
`;

const updateMatrix = (matrix) => {
  const directions = [
    [0, 1], // up
    [0, -1], // down
    [1, 0], // Right
    [-1, 0], // Left
  ];
  const visited = new Map();
  const rows = matrix.length;
  const cols = matrix[0].length;
  const queue = [];
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (matrix[r][c] === 0) {
        queue.push([r, c]);
      }
    }
  }
  console.log(queue);
  while (queue.length) {
    const [r, c] = queue.shift();
    for (let [dr, dc] of directions) {
      const newRow = r + dr;
      const newCol = c + dc;
      if (
        newRow >= 0 &&
        newRow < rows &&
        newCol >= 0 &&
        newCol < cols &&
        matrix[newRow][newCol] === 1 &&
        !visited.get(`${newRow}-${newCol}`)
      ) {
        queue.push([newRow, newCol]);
        visited.set(`${newRow}-${newCol}`, true);
        matrix[newRow][newCol] = matrix[r][c] + 1;
      }
    }
  }
  return matrix;
};
mat = [
  [0, 0, 0],
  [0, 1, 0],
  [1, 1, 1],
];
console.log(updateMatrix(mat));

File: bfs and dfs/542. 01 Matrix.py
"""Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.

The distance between two adjacent cells is 1.

 

Example 1:


Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]"""
from typing import List
from collections import deque

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        rows = len(mat)
        cols = len(mat[0])
        queue = deque()

        # Initialize the queue with 0 cells and mark other cells as infinity
        for i in range(rows):
            for j in range(cols):
                if mat[i][j] == 0:
                    queue.append((i, j))
                else:
                    mat[i][j] = float('inf')

        # Perform BFS traversal starting from 0 cells
        while queue:
            row, col = queue.popleft()
            
            for direction in directions:
                new_row = row + direction[0]
                new_col = col + direction[1]

                # Check if the new position is out of bounds
                if new_row < 0 or new_row >= rows or new_col < 0 or new_col >= cols:
                    continue

                # If the current distance is less than the newly calculated distance,
                # no need to update the cell or enqueue it
                if mat[new_row][new_col] <= mat[row][col] + 1:
                    continue

                # Update the cell with the new distance and enqueue it
                queue.append((new_row, new_col))
                mat[new_row][new_col] = mat[row][col] + 1

        return mat


mat = [[0,0,0],[0,1,0],[1,1,1]]

print(Solution().updateMatrix(mat))
